text/0000-landmarks.md

```diff
@@ -0,0 +1,43 @@
+# Landmark observations
+Add landmarks support to Cartographer.
+
+## Summary
+[summary]: #summary
+Cartographer will be able to work with a new type of sensor data that contains IDs and locations of landmarks.
+
+## Motivation
+[motivation]: #motivation
+Landmarks translation/orientation can improve SLAM quality. Landmarks rendered on the map can be useful for inspection. 
+
+## Approach
+[approach]: #approach
+Landmarks are sparse features of geometry that we want to map. Unlike dense range data, landmarks are assumed to be sparse in comparison. In order to support using them, current SLAM approach has to be generalized. 
+
+### Sensor
+Each landmark will have an ID and a position relative to the robot with or without orientation. A landmark observation can contain several landmarks at a time.
+
+### Submap
+Submaps for 2D and 3D geometries are occupancy grids that contain accumulated range data.  There can be two approaches to add landmarks to the submaps:
+
+1. Submap holds a dictionary with landmark data keyed by landmark IDs;
+2. Every cell of the grid is modified to hold various features, for example, likelihood of landmarks, ML features. 
+
+The second approach is more generic and makes it possible to introduce landmarks based on the strength of the signal, e.g. Wi-Fi signal, BT beacons. The first approach should work faster, it is also more flexible and non-intrusive.
+  
+### Constraint builder
+`ConstraintBuilder` should use landmarks while adding intra-submap constraints:
+
+* `FastCorrelativeScanMatcher` will use landmarks in the bounding function;
+* `CeresScanMatcher` will have an additional term in its cost function.
+
+### Optimization problem
+Landmarks may enter sparse pose graph optimization as additional independent variables. It would increase the dimension of the optimization by `number_of_landmarks * number_of_parameters_per_landmark`, which could be undesirable when there is a significant number of landmarks.
+
+There is another use case here. If Cartographer is provided with the relative positions of landmarks with respect to each other, it can be added to the sparse pose graph optimization similarly to `FixedFramePoseData`.
+
+### Rendering
+Landmarks IDs and positions should be propagated to RViz for rendering.
+
+## Discussion Points
+[discussion]: #discussion
+Should the landmarks be a part of local SLAM if a robot is capable to detect landmarks while moving?
```

Comment:
I can imagine landmark observations to introduce significant jumps that we want to avoid in local SLAM. Think of the case where the local SLAM front-end is used as odometry.

I see the incorporation of landmarks more as a corrective measure to be used in the backend, similar to the way loop closures are utilized.

Reply:
Agreed, especially for the Magazino use case. It would be interesting to experiment with it later though.

Reply:
Sure, but it shouldn't happen implicitly.

Reply:
We know that scan matching already introduces such jumps, which is why the odom topic Cartographer publishes is at times non-smooth and not great for feeding into navstacks.

@MichaelGrupp Are you saying that you expect that landmarks will make those existing jumps bigger?

Reply:
Ok maybe let's forget the odom use case. I mainly struggle with how landmarks that provide global information can be used in local SLAM (which to my understanding should estimate the local motion of the robot).

Let's think of the hypothetical situation where the system observes a landmark that can be associated with an absolute pose inside the global map with high certainty (e.g. some accurately measured marker). While this is certainly beneficial for the optimization to produce a globally consistent map, how should this be incorporated in local SLAM? Imagine we have collected quite a large amount of drift at that point and observe the landmark that tells us we're actually off by a few meters.

Reply:
In local SLAM it would be used exactly like we are using range data now: we get new a new point cloud and optimize translation/rotation w.r.t. the probability grid (2d) / hybrid grid (3d). Assume we have landmark(s) that are observed by robot in every trajectory node inside one of the submaps. Then we would try to maximize f_1(range_data (translation, rotation), grid) + f_2(landmarks_visible_from_trajectory_node, accumulated_landmarks_in_submap), i.e. it would require another criterion in the goal function.

It is definitely not what the use case you want. This would be a more generic approach, but it seems that it is not needed yet. It can be added later if needed. For now, I am removing local SLAM parts from the RFC. I'll expand the global SLAM part tomorrow.

Reply:
Ahh ok, we are probably talking about different things. For me, the first thing that came to my mind was: landmark == semantic point in the environment with known location. The term landmark is a bit overused in SLAM :stuck_out_tongue: 

But looking again at your draft:
> Landmarks are sparse features of geometry that we want to map

So you mean a landmark is a local observation that can be tracked (similar to e.g. corner features in feature-based visual SLAM)? Why not call it features if landmark is just an alias for "sparse features of geometry" to avoid confusion?

Reply:
Thanks, Michael. When I first started writing this, I had an idea of covering all possible landmark use cases(semantic points and sparse features), but then it looks like it is unnecessary at this point. I will concentrate on "semantic points" for now.

Reply:
Ok, thanks for the clarification. Makes also more sense in my opinion. Looking forward to see the final RFC - this definitely something I'd be also interested in.